src/ctx_macro.rs

```diff
@@ -0,0 +1,239 @@
+        let $len = $len_expr;
+        let $offset = $offset_expr;
+        {
+            #[allow(unused_macros)]
+            macro_rules! set {
+                ($buf:expr, $val:expr) => {
+                    assert!($offset <= $buf.len() && $offset + $len <= $buf.len());
+                };
+            }
+            #[allow(unused_macros)]
+            macro_rules! set_disjoint {
+                ($buf:expr, $val:expr) => {
+                    assert!($offset <= $buf.len() && $offset + $len <= $buf.len());
+                };
+            }
+            $body
+        }
+        macro_rules! exec_block {
+            ($N:literal, $block:block) => {
+                {
+                    #[allow(unused_macros)]
+                    macro_rules! set {
+                        ($buf:expr, $val:expr) => {
+                            // SAFETY: The offset and length are checked by the
+                            // assert outside of the match.
+                            let buf_range = unsafe {
+                                $buf.get_unchecked_mut($offset..$offset+$N)
+                            };
+                            *<&mut [_; $N]>::try_from(buf_range).unwrap() = [$val; $N];
+                        };
+                    }
+                    #[allow(unused_macros)]
+                    macro_rules! set_disjoint {
+                        ($buf:expr, $val:expr) => {
+                            // SAFETY: The offset and length are checked by the
+                            // assert outside of the match.
+                            let mut buf_range = unsafe {
+                                $buf.index_mut_unchecked(($offset.., ..$N))
+                            };
+                            *<&mut [_; $N]>::try_from(&mut *buf_range).unwrap() = [$val; $N];
+                        };
+                    }
+                    $block
+                }
+            };
+        }
+        match $len {
+            01 if $UP_TO >= 01 => exec_block!(01, $body),
+            02 if $UP_TO >= 02 => exec_block!(02, $body),
+            04 if $UP_TO >= 04 => exec_block!(04, $body),
+            08 if $UP_TO >= 08 => exec_block!(08, $body),
+            16 if $UP_TO >= 16 => exec_block!(16, $body),
+            32 if $UP_TO >= 32 => exec_block!(32, $body),
+            64 if $UP_TO >= 64 => exec_block!(64, $body),
+            _ => {
+                if $($WITH_DEFAULT ||)? false {
+                    #[allow(unused_macros)]
+                    macro_rules! set {
+                        ($buf:expr, $val:expr) => {
+                            $buf[$offset..][..$len].fill($val);
+                        };
+                    }
+                    #[allow(unused_macros)]
+                    macro_rules! set_disjoint {
+                        ($buf:expr, $val:expr) => {
+                            $buf.index_mut($offset..$offset+$len).fill($val);
+                        };
+                    }
```

Comment:
Should we have the same `*_unchecked()` indexing/`assume`s here, too?

Reply:
:shrug: I'll test it and see if it makes a difference. That should be the slow path anyway.